### PR TITLE
[OpenTelemetry.Extensions] Add baggage Activity processor

### LINKED ISSUE #1650: Request for new component: Baggage Span Processor
# Issue that does not fit into other categories

The Honeycomb distro currently provides a Span processor that takes items from the baggage and adds those items as attributes onto spans. We'd like to donate this code upstream if there's an interest from the community, the code is pretty minimal:

- https://github.com/honeycombio/honeycomb-opentelemetry-dotnet/blob/main/src/Honeycomb.OpenTelemetry/BaggageSpanProcessor.cs

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
Fixes #1650 

## Changes

Adds component - baggage span processor.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [x] Design discussion issue #1650 
* [x] Changes in public API reviewed

### CODE REVIEW

File: .github/component_owners.yml [hunk 1]
```diff
@@ -82,6 +82,8 @@ components:
     - srprash
     - atshaw43
     - ppittle
+  OpenTelemetry.SpanProcessor.Baggage:
```
Comment (cijothomas) [inline .github/component_owners.yml:85]: @MikeGoldsmith @Kielek 
Would this be better made as part of https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Extensions ? Maintaining a new package for a single processor would be bit overkill, I feel!.

File: src/OpenTelemetry.SpanProcessor.Baggage/BaggageSpanProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+using OpenTelemetry;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace OpenTelemetry.SpanProcessor.Baggage;
+
+/// <summary>
+/// Span processor that adds <see cref="Baggage"/> fields to every new span
+/// </summary>
+public class BaggageSpanProcessor : OpenTelemetry.<Activity>
```
Comment (cijothomas) [inline src/OpenTelemetry.SpanProcessor.Baggage/BaggageSpanProcessor.cs:13]: I'd suggest to keep this internal, and offer extension method like https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs#L32

**Comment by MikeGoldsmith** [comment]
Thanks for the feedback @cijothomas - I'll make suggested changed over the next day or two.

File: .github/component_owners.yml (hunk 1 above)
Comment (Kielek) [reply .github/component_owners.yml:85]: I like the idea. Forgot about this package (never need to use it).

**Comment by MikeGoldsmith** [comment]
@Kielek @cijothomas I've updated the PR to add the baggage span processor to the OpenTelemetry.Extensions package instead of creating a new one. Ready for another review 👍🏻

File: src/OpenTelemetry.Extensions/README.md [hunk 3]
```diff
@@ -37,3 +37,33 @@ public static TracerProviderBuilder AddMyExporter(this TracerProviderBuilder bui
         .AddAutoFlushActivityProcessor(a => a.Parent == null && (a.Kind == ActivityKind.Server || a.Kind == ActivityKind.Consumer), 5000);
 }
 ```
+
+### BaggageSpanProcessor
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/README.md:41]: Given .NET already settled on `Activity`, we should name this `BaggageActivityProcessor`, like `AutoFlushActivityProcessor` already part of this package.

File: src/OpenTelemetry.Extensions/README.md [hunk 4]
```diff
@@ -37,3 +37,33 @@ public static TracerProviderBuilder AddMyExporter(this TracerProviderBuilder bui
...
+
+The BaggageSpanProcessor reads entries stored in Baggage from the parent context
+and adds the baggage keys and values to the activity as attributes on start.
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/README.md:44]:
```suggestion
and adds the baggage keys and values to the `Activity` as tags(attributes) on start.
```

File: src/OpenTelemetry.Extensions/README.md [hunk 5]
```diff
@@ -37,3 +37,33 @@ public static TracerProviderBuilder AddMyExporter(this TracerProviderBuilder bui
...
+
+Add this span processor to a tracer provider.
+
+Warning!
+
+Do not put sensitive information in Baggage.
+
+To repeat: a consequence of adding data to Baggage is that the keys and values
+will appear in all outgoing HTTP headers from the application.
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/README.md:53]: could we make this bit more generic? HTTP headers is one common way, but not the only way.

File: src/OpenTelemetry.Extensions/README.md [hunk 6]
```diff
@@ -37,3 +37,33 @@ public static TracerProviderBuilder AddMyExporter(this TracerProviderBuilder bui
...
+### BaggageSpanProcessor
+
+The BaggageSpanProcessor reads entries stored in Baggage from the parent context
+and adds the baggage keys and values to the activity as attributes on start.
+
+Add this span processor to a tracer provider.
+
+Warning!
+
+Do not put sensitive information in Baggage.
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/README.md:50]: mm..I actually think this warning belongs to the Baggage documentation, not this, right? This doc is merely copying values already in Baggage and putting it as Activity Tags.

File: src/OpenTelemetry.Extensions/README.md [hunk 7]
```diff
@@ -37,3 +37,33 @@ public static TracerProviderBuilder AddMyExporter(this TracerProviderBuilder bui
...
+
+To repeat: a consequence of adding data to Baggage is that the keys and values
+will appear in all outgoing HTTP headers from the application.
+
+Example of AddBaggageSpanProcessor:
+
+```cs
+public static TracerProviderBuilder AddMyExporter(this TracerProviderBuilder builder, MyExporterOptions options)
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/README.md:58]: I think this is a copy-paste error as it is showing Exporter!

File: .github/component_owners.yml [hunk 8]
```diff
@@ -20,6 +20,7 @@ components:
     - SergeyKanzhelev
   src/OpenTelemetry.Extensions/:
     - codeblanch
+    - mikegoldsmith
```
Comment (Kielek) [inline .github/component_owners.yml:23]: Please keep in sync with test/OpenTelemetry.Extensions.Tests/

@CodeBlanch, FYI

**Comment by MikeGoldsmith** [comment]
@cijothomas @Kielek updated as requested, let me know if you have any more feedback.

File: src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs [hunk 9]
```diff
@@ -45,6 +45,28 @@ public static class TracerProviderBuilderExtensions
...
+    /// <summary>
+    /// Adds the <see cref="BaggageActivityProcessor"/> to the <see cref="TracerProviderBuilder"/>.
+    /// </summary>
+    /// <param name="builder"><see cref="TracerProviderBuilder"/> to add the <see cref="BaggageActivityProcessor"/> to.</param>
+    /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
+    public static TracerProviderBuilder AddBaggageActivityProcessor(
+        this TracerProviderBuilder builder)
+    {
+#if NET6_0_OR_GREATER
+        ArgumentNullException.ThrowIfNull(builder);
```
Comment (Kielek) [inline src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs:60]: Please check `Guard` class in this repo. It is available also for .NET Framework.

**Review by cijothomas — APPROVED** [review]
Looks good. 
Left couple of non-blocking comments.

File: src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs (hunk 9 above)
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs]: Added in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1659/commits/2a4852ba4bf30e10d36c875393b4bb243ccf52c6

File: src/OpenTelemetry.Extensions/README.md [hunk 10]
```diff
@@ -37,3 +37,25 @@ public static TracerProviderBuilder AddMyExporter(this TracerProviderBuilder bui
...
+The BaggageActivityProcessor reads entries stored in Baggage from the parent context
+and adds the baggage keys and values to the `Activity` as tags(attributes) on start.
+
+Add this activity processor to a tracer provider.
+
+Warning!
+
+Do not put sensitive information in Baggage.
+
+To repeat: a consequence of adding data to Baggage is that the keys and values
```
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/README.md:52]: I don't think I understand. That's the intention of the baggage span processor; baggage entries are added to all subsequent Activities.

**Comment by martinjt** [comment]
I'd prefer if we had a constructor that allowed the user to add a prefix for baggage items that they'd like to add as attributes. We've had situations where customers have used a similar approach and managed to receive data from externally propagated baggage (accidentally).

A pattern like adding a prefix would mean that it becomes a lot more intentional, and therefore just a little safer.

**Comment by MikeGoldsmith** [comment]
> I'd prefer if we had a constructor that allowed the user to add a prefix for baggage items that they'd like to add as attributes. We've had situations where customers have used a similar approach and managed to receive data from externally propagated baggage (accidentally).
> 
> A pattern like adding a prefix would mean that it becomes a lot more intentional, and therefore just a little safer.

I'd be happy to discuss this in a follow-up issue and/or PR. I'd prefer not to complicate the current PR as the suggested change is purely additive that someone can opt-in to using if desired.

**Comment by martinjt** [comment]
> > I'd prefer if we had a constructor that allowed the user to add a prefix for baggage items that they'd like to add as attributes. We've had situations where customers have used a similar approach and managed to receive data from externally propagated baggage (accidentally).
> > A pattern like adding a prefix would mean that it becomes a lot more intentional, and therefore just a little safer.
> 
> I'd be happy to discuss this in a follow-up issue and/or PR. I'd prefer not to complicate the current PR as the suggested change is purely additive that someone can opt-in to using if desired.

I think the fact that we're adding this to contrib gives this a bit more credibility, and therefore getting it right first time is more important. Putting this in as-is can lead to issues and therefore I think it's better to produce this in the way that makes it safest for customers.

**Comment by MikeGoldsmith** [comment]
> > > I'd prefer if we had a constructor that allowed the user to add a prefix for baggage items that they'd like to add as attributes. We've had situations where customers have used a similar approach and managed to receive data from externally propagated baggage (accidentally).
> > > A pattern like adding a prefix would mean that it becomes a lot more intentional, and therefore just a little safer.
> > 
> > 
> > I'd be happy to discuss this in a follow-up issue and/or PR. I'd prefer not to complicate the current PR as the suggested change is purely additive that someone can opt-in to using if desired.
> 
> I think the fact that we're adding this to contrib gives this a bit more credibility, and therefore getting it right first time is more important. Putting this in as-is can lead to issues and therefore I think it's better to produce this in the way that makes it safest for customers.

How would you like to see prefixes implemented, Include or exclude list? Should it handle whole matches or just prefixes? Both?

I don't think it's a simple answer and could get complicated quickly. This is the reason why I added warnings to both the README and class description but was asked to remove as it was deemed unnecessary.

I think adding something that's simple and clearly explains the pitfalls and chose to make incremental improvements is positive for this project. Attempting to make it perfect from outset is discouraging community participation. The saying of "perfection is the enemy of progress" could be true here.

**Comment by MikeGoldsmith** [comment]
PS we (Honeycomb) have opened PRs to contribute the same baggage span (activity) processor to a number of SDK contrib projects.i feel altering the behaviour of one could be disjointing for any organisation that uses multiple languages
- https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/937
- https://github.com/open-telemetry/opentelemetry-js-contrib/pull/2114
- https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5404
- https://github.com/open-telemetry/opentelemetry-java-contrib/pull/1290
- https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2436

**Comment by martinjt** [comment]
There's also the saying "Help people into the pit of success", which we're not doing here.

I'd say rushing this out is not the best option for that. We should workout the best approach, perhaps not in a PR, but an issue.

To me, adding a prefix is the best option.

I appreciate that this has been added for others, but that doesn't make it the right approach, just that there are more approaches that we've added. To be clear this has been an issue in the honeycomb distro's for a long time as you know.

File: src/OpenTelemetry.Extensions/README.md [hunk 11]
```diff
@@ -37,3 +37,18 @@ public static TracerProviderBuilder AddMyExporter(this TracerProviderBuilder bui
         .AddAutoFlushActivityProcessor(a => a.Parent == null && (a.Kind == ActivityKind.Server || a.Kind == ActivityKind.Consumer), 5000);
 }
 ```
+
+### BaggageActivityProcessor
+
+The BaggageActivityProcessor reads entries stored in Baggage from the current `Baggage`
+and adds the baggage keys and values to the `Activity` as tags(attributes) on start.
```
Comment (Kielek) [inline src/OpenTelemetry.Extensions/README.md:44]:
```suggestion
and adds the baggage keys and values to the `Activity` as tags (attributes) on start.
```

**Comment by MikeGoldsmith** [comment]
To be clear; adding sensitive data to baggage is not a good idea, regardless of using this processor. I don't disagree the security concern is real but don't think it's a problem with this processor. I think trace context should have a separate bucket of shared data that is separate from baggage.

@Kielek I've opened the following issue so we can continue the discussion outside of this PR:
- https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1695

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace OpenTelemetry.Trace;
+
+/// <summary>
+/// Activity processor that adds <see cref="Baggage"/> fields to every new span.
+/// </summary>
+internal sealed class BaggageActivityProcessor : BaseProcessor<Activity>
+{
+    /// <inheritdoc />
+    public override void OnStart(Activity activity)
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:14]: @MikeGoldsmith Should we check `Activity.IsAllDataRequested` before adding the tags?

Comment (cijothomas) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:14]: Not needed right? processors won't be called at all without that

**Pull request merged**